queries/places.py

```diff
@@ -0,0 +1,352 @@
+    result["qkey"] = "OpeningHours"
+
+
+def QPlacesIsOpen(node, params, result):
+    result["qkey"] = "IsOpen"
+
+
+def QPlacesIsClosed(node, params, result):
+    result["qkey"] = "IsClosed"
+
+
+def QPlacesAddress(node, params, result):
+    result["qkey"] = "PlaceAddress"
+
+
+def QPlacesSubjectNf(node, params, result):
+    result["subject_nom"] = _fix_placename(result._nominative)
+
+
+QPlacesSubjectEf = QPlacesSubjectÞgf = QPlacesSubjectÞf = QPlacesSubjectNf
+
+
+_PLACES_API_ERRMSG = "Ekki tókst að fletta upp viðkomandi stað"
+_NOT_ICELAND_ERRMSG = "Enginn staður með þetta heiti fannst á Íslandi"
+
+
+def _parse_coords(place):
+    try:
+        lat = float(place["geometry"]["location"]["lat"])
+        lng = float(place["geometry"]["location"]["lng"])
+        return (lat, lng)
+    except Exception as e:
+        logging.warning(
+            "Unable to parse place coords for place {0}: {1}".format(place, e)
+        )
+    return None
+
+
+def answ_address(placename, loc, qtype):
+    # Look up placename in places API
+    res = query_places_api(
+        placename, userloc=loc, fields="formatted_address,name,geometry"
+    )
+
+    if res["status"] != "OK" or "candidates" not in res or not res["candidates"]:
+        return gen_answer(_PLACES_API_ERRMSG)
+
+    # Use top result
+    place = res["candidates"][0]
+
+    # Make sure it's in Iceland
+    coords = _parse_coords(place)
+    if not coords or not in_iceland(coords):
+        return gen_answer(_NOT_ICELAND_ERRMSG)
+
+    # Remove superfluous "Ísland" in addr string
+    addr = re.sub(r", Ísland$", "", place["formatted_address"])
+    # Get street name without number to get preposition
+    street_name = addr.split()[0].rstrip(",")
+    prep = iceprep_for_street(street_name)
+    # Split addr into street name w. number, and remainder
+    street_addr = addr.split(",")[0]
+    remaining = re.sub(r"^{0}".format(street_addr), "", addr)
+    # Get street name in dative case
+    addr_þgf = NounPhrase(street_addr).dative
+    # Assemble final address
+    final_addr = "{0}{1}".format(addr_þgf, remaining)
+
+    # Create answer
+    answer = final_addr
+    voice = "{0} er {1} {2}".format(placename, prep, numbers_to_neutral(final_addr))
+    response = dict(answer=answer)
+
+    return response, answer, voice
+
+
+def answ_openhours(placename, loc, qtype):
+    # Look up placename in places API
+    res = query_places_api(
+        placename,
+        userloc=loc,
+        fields="opening_hours,place_id,formatted_address,geometry",
+    )
+    if res["status"] != "OK" or "candidates" not in res or not res["candidates"]:
+        return gen_answer(_PLACES_API_ERRMSG)
+
+    from pprint import pprint
+    pprint(res)
+
+    # Use top result
+    place = res["candidates"][0]
+    if not "opening_hours" in place:
+        return gen_answer("Ekki tókst að sækja opnunartíma fyrir " + icequote(placename))
+
+    place_id = place["place_id"]
+    is_open = place["opening_hours"]["open_now"]
```

Comment:
If we are caching this, be sure that it is not for too long! (Maybe 5 minutes max.)

Reply:
We are not caching this at the moment. But it would probably be more efficient to do so, and compute whether the places is open based on opening hours and current time. Will do so in next iteration.